sdk/messaging/azservicebus/internal/amqpLinks.go

```diff
@@ -97,240 +101,238 @@ type CreateLinkFunc func(ctx context.Context, session AMQPSession) (AMQPSenderCl
-	}
-
-	return err
-}
-
-func (links *amqpLinks) recoverImpl(ctx context.Context, try int, linksRevision uint64, origErr error) error {
-	_, span := tab.StartSpan(ctx, tracing.SpanRecoverLink)
-	defer span.End()
-
 	if origErr == nil || IsCancelError(origErr) {
 		return nil
 	}
 
+	log.Writef(EventConn, "Recovering link for error %s", origErr.Error())
+
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	default:
 	}
 
-	span.AddAttributes(tab.Int64Attribute("attempt", int64(try)))
-
-	if shouldRecreateLink(origErr) {
-		span.AddAttributes(
-			tab.StringAttribute("recovery_kind", "link"),
-			tab.StringAttribute("error", origErr.Error()),
-			tab.StringAttribute("error_type", fmt.Sprintf("%T", origErr)))
+	sbe := GetSBErrInfo(origErr)
 
-		if err := links.recoverLink(ctx, &linksRevision); err != nil {
-			span.AddAttributes(tab.StringAttribute("recoveryFailure", err.Error()))
+	if sbe.RecoveryKind == RecoveryKindLink {
+		if err := links.recoverLink(ctx, theirID); err != nil {
+			log.Writef(EventConn, "failed to recreate link: %s", err.Error())
 			return err
 		}
 
+		log.Writef(EventConn, "Recovered links")
 		return nil
-	} else if shouldRecreateConnection(ctx, origErr) {
-		span.AddAttributes(
-			tab.StringAttribute("recovery_kind", "connection"),
-			tab.StringAttribute("error", origErr.Error()),
-			tab.StringAttribute("error_type", fmt.Sprintf("%T", origErr)))
-
-		if err := links.recoverConnection(ctx); err != nil {
-			span.Logger().Error(fmt.Errorf("failed to recreate connection: %w", err))
-			return err
-		}
-
-		// unconditionally recover the link if the connection died.
-		if err := links.recoverLink(ctx, nil); err != nil {
-			span.Logger().Error(fmt.Errorf("failed to recover links after connection restarted: %w", err))
+	} else if sbe.RecoveryKind == RecoveryKindConn {
+		if err := links.recoverConnection(ctx, theirID); err != nil {
+			log.Writef(EventConn, "failed to recreate connection: %s", err.Error())
 			return err
 		}
 
+		log.Writef(EventConn, "Recovered connection and links")
 		return nil
 	}
 
-	span.AddAttributes(
-		tab.StringAttribute("recovery", "none"),
-		tab.StringAttribute("error", origErr.Error()),
-		tab.StringAttribute("errorType", fmt.Sprintf("%T", origErr)))
-
+	log.Writef(EventConn, "Recovered, no action needed")
 	return nil
 }
 
-func (links *amqpLinks) recoverConnection(ctx context.Context) error {
+func (links *AMQPLinksImpl) recoverConnection(ctx context.Context, theirID LinkID) error {
 	tab.For(ctx).Info("Connection is dead, recovering")
 
-	links.mu.RLock()
-	clientRevision := links.clientRevision
-	links.mu.RUnlock()
+	links.mu.Lock()
+	defer links.mu.Unlock()
 
-	err := links.ns.Recover(ctx, clientRevision)
+	created, err := links.ns.Recover(ctx, uint64(theirID.Conn))
 
 	if err != nil {
-		tab.For(ctx).Error(fmt.Errorf("Recover connection failure: %w", err))
+		log.Writef(EventConn, "Recover connection failure: %s", err)
 		return err
 	}
 
+	if created || theirID.Link == links.id.Link {
```

Comment:
Basically, recovery here is:
- If we recreated the connection we _definitely_ need to recreate our link
- If the link we got an error on matches our current link we need to recreate our link. (if the ID _didn't_ match then that indicates we already recovered, so they just need to re-get the link).

Reply:
This explanation would probably be a good code comment here.

Reply:
Yeah, was thinking the same thing when I wrote it out. :) Will add.

Reply:
Added some more comments on this and the function it calls.